lib/model/losses.py

```diff
@@ -509,46 +481,52 @@ def scharr_edges(image, magnitude):
-                [-2.0, -2.0, 0.0, 2.0, 2.0],
-                [-3.0, -6.0, 0.0, 6.0, 3.0],
-                [-2.0, -2.0, 0.0, 2.0, 2.0],
-                [-1.0, -1.0, 0.0, 1.0, 1.0]]]
-    num_kernels = len(kernels)
-    kernels = np.transpose(np.asarray(kernels), (1, 2, 0))
-    kernels = np.expand_dims(kernels, -2) / np.sum(np.abs(kernels))
-    kernels_tf = tf.constant(kernels, dtype=image.dtype)
-    kernels_tf = tf.tile(kernels_tf, [1, 1, image_shape[-1], 1], name='scharr_filters')
+    image_shape = K.shape(image)
+
+    # 5x5 modified Scharr kernel ( reshape to (5,5,1,2) )
+    matrix = [[[[0.00070, 0.00070]],
+               [[0.00520, 0.00370]],
+               [[0.03700, 0.00000]],
+               [[0.00520, -0.0037]],
+               [[0.00070, -0.0007]]],
+              [[[0.00370, 0.00520]],
+               [[0.11870, 0.11870]],
+               [[0.25890, 0.00000]],
+               [[0.11870, -0.1187]],
+               [[0.00370, -0.0052]]],
+              [[[0.00000, 0.03700]],
+               [[0.00000, 0.25890]],
+               [[0.00000, 0.00000]],
+               [[0.00000, -0.2589]],
+               [[0.00000, -0.0370]]],
+              [[[-0.0037, 0.00520]],
+               [[-0.1187, 0.11870]],
+               [[-0.2589, 0.00000]],
+               [[-0.1187, -0.1187]],
+               [[-0.0037, -0.0052]]],
+              [[[-0.0007, 0.00070]],
+               [[-0.0052, 0.00370]],
+               [[-0.0370, 0.00000]],
+               [[-0.0052, -0.0037]],
+               [[-0.0007, -0.0007]]]]
+    num_kernels = [2]
+    kernels = K.constant(matrix, dtype='float32')
+    kernels = K.tile(kernels, [1, 1, image_shape[-1], 1])
+
 
     # Use depth-wise convolution to calculate edge maps per channel.
+    # Output tensor has shape [batch_size, h, w, d * num_kernels].
     pad_sizes = [[0, 0], [2, 2], [2, 2], [0, 0]]
     padded = tf.pad(image, pad_sizes, mode='REFLECT')
+    output = K.depthwise_conv2d(padded, kernels)
 
-    # Output tensor has shape [batch_size, h, w, d * num_kernels].
-    strides = [1, 1, 1, 1]
-    output = tf.nn.depthwise_conv2d(padded, kernels_tf, strides, 'VALID')
-
-    # Reshape to [batch_size, h, w, d, num_kernels].
-    shape = tf.concat([image_shape, [num_kernels]], 0)
-    output = tf.reshape(output, shape=shape)
-    output.set_shape(static_image_shape.concatenate([num_kernels]))
-
-    if magnitude:  # magnitude of edges
-        output = tf.sqrt(tf.reduce_sum(tf.square(output), axis=-1))
-    else:  # direction of edges
-        output = tf.atan(tf.squeeze(tf.div(output[:, :, :, :, 0]/output[:, :, :, :, 1])))
+    if not magnitude:  # direction of edges
+        # Reshape to [batch_size, h, w, d, num_kernels].
+        shape = K.concatenate([image_shape, num_kernels], axis=0)
+        output = K.reshape(output, shape=shape)
+        output.set_shape(static_image_shape.concatenate(num_kernels))
+        output = tf.atan(K.squeeze(output[:, :, :, :, 0] / output[:, :, :, :, 1]))
```

Comment:
Not for myself. Add `plaidml.tile.unary_op(K.ones(10), "atan(I)")` as plaidml fallback.